src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssemblyOrExcuse.java

```diff
@@ -111,41 +114,80 @@ public FermiLiteAssembly getAssembly() {
     /**
      * write a SAM file containing records for each aligned contig
      */
-    public static void writeSAMFile( final String samFile, final PipelineOptions pOpts,
-                                     final SAMFileHeader header,
-                                     final List<AlignedAssemblyOrExcuse> alignedAssemblyOrExcuseList ) {
+    static void writeSAMFile( final String samFile, final PipelineOptions pOpts,
+                              final SAMFileHeader header,
+                              final List<AlignedAssemblyOrExcuse> alignedAssemblyOrExcuseList ) {
         try ( final OutputStream os = BucketUtils.createFile(samFile, pOpts) ) {
             final SAMTextWriter writer = new SAMTextWriter(os);
             writer.setSortOrder(SAMFileHeader.SortOrder.queryname, true);
             writer.setHeader(header);
-            final List<String> refNames =
-                    header.getSequenceDictionary().getSequences().stream()
-                            .map(SAMSequenceRecord::getSequenceName).collect(Collectors.toList());
-            for ( final AlignedAssemblyOrExcuse alignedAssemblyOrExcuse : alignedAssemblyOrExcuseList ) {
-                if ( alignedAssemblyOrExcuse.getErrorMessage() != null ) continue;
-                final FermiLiteAssembly assembly = alignedAssemblyOrExcuse.getAssembly();
-                final int assemblyId = alignedAssemblyOrExcuse.getAssemblyId();
-                final List<List<BwaMemAlignment>> allAlignments = alignedAssemblyOrExcuse.getContigAlignments();
-                final int nContigs = assembly.getNContigs();
-                for ( int contigIdx = 0; contigIdx != nContigs; ++contigIdx ) {
-                    final List<BwaMemAlignment> alignments = allAlignments.get(contigIdx);
-                    if ( alignments.isEmpty() ) continue;
-                    final FermiLiteAssembly.Contig contig = assembly.getContig(contigIdx);
-                    final String readName = String.format("asm%06d:tig%05d", assemblyId, contigIdx);
-                    final byte[] calls = contig.getSequence();
-                    for ( final BwaMemAlignment alignment : alignments ) {
-                        writer.addAlignment(BwaMemAlignmentUtils.applyAlignment(readName, calls, null, null, alignment,
-                                refNames, header, false, false));
-                    }
-                }
-            }
+            filterFailedAssemblyAndConvertToSamRecords(header, alignedAssemblyOrExcuseList).forEach(contigs -> contigs.forEach(ls -> ls.forEach(writer::addAlignment)) );
```

Comment:
This line is too long.  Our style guide says 100 for max line length.  But I'd be fine with the IntelliJ default of 120.

Personally, I find the functional-style breakout to be harder to read, and more obscure than the original code, but I won't object on that score.  *De gustibus non est disputandem.*  I agree that the original code -- with three nested for loops -- was too dense and complex.
What I will object to is gathering a List<List<List<SAMRecord>>> and then unfolding it all in this line.  What should be done instead, IMHO, is to start with the alignedAssemblyOrExcuseList, and then...
```
.filter( failed assemblies )
.map( to a stream of assemblies )
.flatmap( to a stream of contigs )
.flatmap( to a stream of alignments )
.map( each alignment to a SAMRecord )
.foreach( add to the writer )
```
The original code required virtually no extra working memory (no temporary collections, anyway).
Your revision requires everything to live in memory all at once before any writing is done.
The revised functional version sketched above requires minimal extra working memory.